zebra-network/src/peer_set/set/tests/vectors.rs

```diff
@@ -0,0 +1,161 @@
+                assert_eq!(client_request.request, Request::Peers)
+            }
+            _ => unreachable!(),
+        };
+
+        // Drop the future
+        std::mem::drop(fut);
+
+        // Peer set will remain ready for requests
+        let peer_ready2 = peer_set
+            .ready()
+            .await
+            .expect("peer set service is always ready");
+
+        // Get a new future calling a different request than before
+        let _fut = peer_ready2.call(Request::MempoolTransactionIds);
+
+        // Client received the request
+        match client_handle.try_to_receive_outbound_client_request() {
+            ReceiveRequestAttempt::Request(client_request) => {
+                assert_eq!(client_request.request, Request::MempoolTransactionIds)
+            }
+            _ => unreachable!(),
+        };
+    });
+}
+
+#[test]
+fn peer_set_ready_multiple_connections() {
+    // Use three peers with the same version
+    let peer_version = Version::min_specified_for_upgrade(Network::Mainnet, NetworkUpgrade::Canopy);
+    let peer_versions = PeerVersions {
+        peer_versions: vec![peer_version, peer_version, peer_version],
+    };
+
+    // Start the runtime
+    let runtime = zebra_test::init_async();
+    let _guard = runtime.enter();
+
+    // Get peers and client handles of them
+    let (discovered_peers, handles) = peer_versions.mock_peer_discovery();
+    let (minimum_peer_version, _best_tip_height) =
+        MinimumPeerVersion::with_mock_chain_tip(Network::Mainnet);
+
+    // Make sure we have the right number of peers
+    assert_eq!(handles.len(), 3);
+
+    // Build a peerset
+    runtime.block_on(async move {
+        let (mut peer_set, _peer_set_guard) = PeerSetBuilder::new()
+            .with_discover(discovered_peers)
+            .with_minimum_peer_version(minimum_peer_version.clone())
+            .build();
+
+        // Get peerset ready
+        let peer_ready = peer_set
+            .ready()
+            .await
+            .expect("peer set service is always ready");
+
+        // Check we have the right amount of ready services
+        assert_eq!(peer_ready.ready_services.len(), 3);
+
+        // Stop some peer connections but not all
+        handles[0].stop_connection_task().await;
+        handles[1].stop_connection_task().await;
+
+        // We can still make the peer set ready
+        let peer_ready = peer_set
+            .ready()
+            .await
+            .expect("peer set service is always ready");
+
+        // We should have less ready peers
+        assert_eq!(peer_ready.ready_services.len(), 1);
+
+        // Stop the connection of the last peer
+        handles[2].stop_connection_task().await;
+
+        // Peer set hangs when no more connections are present
+        let peer_ready = peer_set.ready();
+        assert!(timeout(Duration::from_millis(10), peer_ready)
+            .await
+            .is_err());
```

Comment:
Optional: I'd suggest increasing this timeout as an extra precaution against any regressions that could make the service become ready incorrectly after a while:

```suggestion
        assert!(timeout(Duration::from_secs(10), peer_ready).await.is_err());
```

The test will still be really fast if you use a paused runtime timer. See the other suggestion for this.